tools/src/vcf2genomicsdb_init.cc

```diff
@@ -0,0 +1,993 @@
+
+  // Write out loader.json
+  json.clear();
+  google::protobuf::util::MessageToJsonString(*import_config_protobuf, &json, json_options);
+  g_logger.debug("Writing out loader json file to {}", loader_output);
+  fix_protobuf_generated_json_for_int64(json, {"tiledb_column", "size_per_column_partition", "segment_size",
+                                               "num_cells_per_tile"});
+  TileDBUtils::write_file(loader_output, json.data(), json.length());
+  delete import_config_protobuf;
+
+  return 0;
+}
+
+static int rename_file(const std::string& src, const std::string& dest) {
+  if (TileDBUtils::is_file(dest)) {
+    if (TileDBUtils::delete_file(dest)) {
+      g_logger.error("Could not delete existing file at path {} for renaming {}", dest, src);
+      return ERR;
+    }
+  }
+  if (TileDBUtils::move_across_filesystems(src, dest)) {
+     g_logger.error("Could not rename file {} to {}", src, dest);
+     return ERR;
+  }
+  if (TileDBUtils::delete_file(src)) {
+    g_logger.error("Could not delete {} after copying to {}", src, dest);
+    return ERR;
+  }
+  return OK;
+}
+
+static int update_json(import_config_t import_config) {
+  char *callset_contents;
+  size_t callset_length;
+  if (TileDBUtils::read_entire_file(import_config.callset_output, (void **)&callset_contents, &callset_length)) {
+    g_logger.error("Could not read callset json file {}", import_config.callset_output);
+    return ERR;
+  }
+  google::protobuf::StringPiece callset_string(callset_contents, callset_length);
+  google::protobuf::util::JsonParseOptions json_options;
+  json_options.ignore_unknown_fields = true;
+
+  CallsetMappingPB* callset_protobuf = new CallsetMappingPB();
+  auto status = google::protobuf::util::JsonStringToMessage(callset_string, callset_protobuf, json_options);
+  if (!status.ok()) {
+    g_logger.error("Protobuf could not apply {} to CallsetMappingPB {}", import_config.callset_output, status.message().as_string());
+    free(callset_contents);
+    return ERR;
+  }
+  free(callset_contents);
+
+  auto existing_callset_size = callset_protobuf->callsets_size();
+  std::vector<std::string> existing_samples;
+  for (auto callset:callset_protobuf->callsets()) {
+    existing_samples.push_back(callset.sample_name());
+  }
+  auto row_index = callset_protobuf->callsets()[existing_callset_size-1].row_idx() + 1;
+
+  bool found_new_samples = false;
+  for (auto sample_uri: import_config.samples) {
+    genomicsdb_htslib_plugin_initialize(sample_uri.c_str());
+    htsFile* fptr = hts_open(sample_uri.c_str(), "r");
+    if (!fptr) {
+      g_logger.error("Could not open sample {} with hts_open {}", sample_uri, strerror(errno));
+      return ERR;
+    }
+    bcf_hdr_t* hdr = bcf_hdr_read(fptr);
+    if (!hdr) {
+      g_logger.error("Could not read sample {} with htslib", sample_uri);
+      return ERR;
+    }
+    for (auto i=0; i<bcf_hdr_nsamples(hdr); i++) {
+      // Validate against existing callsets
+      bool validated = true;
```

Comment:
I wonder if it is worth considering using a map or something else more efficient here for `existing_samples`. As the number of samples goes up we may end up needing a lot of iterations here.

Reply:
Yeah, moving to std::set which uses binary search to check.